include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h

```diff
@@ -412,68 +412,71 @@ using __repacked_tuple_t = typename __repacked_tuple<T>::type;
-// future and helper classes for async pattern/algorithm
-//-----------------------------------------------------------------------
+template <typename _Event>
+void __wait_event(_Event);
 
-// TODO: towards higher abstraction and generic future. implementation specific sycl::event should be hidden
-struct __future_base
+void
+__wait_event(sycl::event __e)
 {
-    sycl::event __my_event;
-
-    __future_base() = default;
-    __future_base(sycl::event __e) : __my_event(__e) {}
-    void
-    wait()
-    {
 #if !ONEDPL_ALLOW_DEFERRED_WAITING
-        __my_event.wait_and_throw();
+    __e.wait_and_throw();
 #endif
-    }
-    operator sycl::event() const { return __my_event; }
-};
+}
 
 template <typename _T>
-class __future : public __future_base
+constexpr auto
+__get_value(sycl::buffer<_T>& __buf)
 {
-    ::std::size_t __result_idx;
-    sycl::buffer<_T> __data;
-
-  public:
-    __future(sycl::event __e, size_t __o, sycl::buffer<_T> __b)
-        : __par_backend_hetero::__future_base(__e), __result_idx(__o), __data(__b)
-    {
-    }
+    //according to a contract, returned value is one-element sycl::buffer
+    return __buf.template get_access<access_mode::read>()[0];
+}
 
-    _T
-    get()
-    {
-        return __data.template get_access<access_mode::read>()[__result_idx];
-    }
-    template <class _Tp, class _Enable>
-    friend class oneapi::dpl::__internal::__future;
-};
+template <typename _T>
+constexpr auto
+__get_value(_T& __val)
+{
+    return __val;
+}
 
-template <>
-class __future<void> : public __future_base
+//A contract for future class: <sycl::event or other event, a value or sycl::buffers...>
+//Impl details: inheretance (private) instead of aggregation for enabling the empty base optimization.
+template <typename _Event, typename... Args>
+class __future : private std::tuple<Args...>
 {
-    ::std::unique_ptr<oneapi::dpl::__internal::__lifetime_keeper_base> __tmps;
+    _Event __my_event;
+    static constexpr bool __is_value = sizeof...(Args) > 0;
 
   public:
-    template <typename... _Ts>
-    __future(sycl::event __e, _Ts... __t) : __future_base(__e)
+    __future(_Event e, Args... args) : __my_event(e), std::tuple<Args...>(args...) {}
+    __future(_Event e, std::tuple<Args...> _t) : __my_event(e), std::tuple<Args...>(_t) {}
+
+    operator _Event() const { return __my_event; }
+    auto
+    event() const
     {
-        if (sizeof...(__t) != 0)
-            __tmps = ::std::unique_ptr<oneapi::dpl::__internal::__lifetime_keeper<_Ts...>>(
-                new oneapi::dpl::__internal::__lifetime_keeper<_Ts...>(__t...));
+        return operator _Event();
     }
     void
+    wait()
+    {
+        __wait_event(operator _Event());
+    }
+
+    template <typename = std::enable_if<__is_value, void>>
+    auto
     get()
     {
-        this->wait();
+        return __get_value(std::get<0>(*this));
+    }
+
+    template <typename _T>
+    auto
+    add_value(_T __t) const
```

Comment:
I don't understand what this method does. Is it public API? If so, let's carefully review what we want to be public API, if not, let's somehow make a clear border between public API and the API for internal use. For example: `__add_value` looks good enough to me.

Reply:
I do not like this to be a method, because semantically it looks like a modifier of the current future object, while in fact it constructs a new object of a different type. I prefer such a conversion to be clearly seen in the code. Can we make a CTAD constructor work for this? The use would look like this:

    return  __future{ value_to_add. another_future }; // template parameters are deduced from ctor arguments

If that will not work, let's create a "make" function instead of a method; the name can be chosen to tell what's going on (for example, `__make_prepend_future(value_to_add. another_future)`)

Reply:
It is the internal API. There are cases where the implementation specify return value  "higher" than SYCL backend, where a future is created. I agree with special name `__add_value`.

Reply:
Actually, I could not define a such constructor `__future{ value_to_add, another_future }`, because class layout is changed...
I think `__make_future(value_to_add) ` should be  OK.